tests/python/relay/test_pass_instrument.py

```diff
@@ -168,3 +169,329 @@ def run_after_pass(self, mod, info):
     # Out of pass context scope, should be reset
     assert passes_counter.run_before_count == 0
     assert passes_counter.run_after_count == 0
+
+
+def test_enter_pass_ctx_exception(capsys):
+    @pass_instrument
+    class PI:
+        def __init__(self, id):
+            self.id = id
+
+        def enter_pass_ctx(self):
+            print(self.id + " enter ctx")
+
+        def exit_pass_ctx(self):
+            print(self.id + " exit ctx")
+
+    @pass_instrument
+    class PIBroken(PI):
+        def __init__(self, id):
+            super().__init__(id)
+
+        def enter_pass_ctx(self):
+            print(self.id + " enter ctx")
+            raise RuntimeError("Just a dummy error")
+
+    pass_ctx = tvm.transform.PassContext(instruments=[PI("%1"), PIBroken("%2"), PI("%3")])
+    with pytest.raises(tvm.error.TVMError):
+        with pass_ctx:
+            pass
+
+    assert "%1 enter ctx\n" "%2 enter ctx\n" == capsys.readouterr().out
```

Comment:
should this also contain "%1 exit ctx\n"?

Reply:
I think no, since `__enter__` didn't catch the exceptions if `EnterPassContent()` throws ([transform.py:89](https://github.com/apache/tvm/blob/6f82e98aac18619078792661d151f33e37afa94d/python/tvm/ir/transform.py#L89)).

Reply:
hrm...I guess this one has no analogue in LLVM. but I'm not sure I agree--I think if an `__enter__` returns succesfully, the contract is that it's expected the `__exit__` will get called. i don't think you can implement `__enter__` with knowledge of any following `__enter__`, so this case would potentially leave state initialized...

since catching `std::runtime_error` is how we handle errors at the FFI, it seems like we should be able to catch that here and invoke the `__exit__` in reverse order. on the other hand, this gets messy, since `__exit__` can also throw exceptions. @tqchen maybe you have some thoughts on doing such complex stuff inside `libtvm_runtime`?

Reply:
I am not quite understanding the sentence - i don't think you can implement `__enter__` with knowledge of any following `__enter___`.

I think ```__enter__``` and ```__exit__``` throw exception is a common problem. Maybe use another RFC/PR to address this?

Reply:
I  considered to catch all the exceptions in `InstrumentEnterPassContext` before:
   if there is exception, disabling instruments within this context, and **NO** throw out to higher level.

But I finally throw it out, since I'd like to see the errors, and don't want to proceed with unstable states.

Reply:
> I am not quite understanding the sentence - i don't think you can implement __enter__ with knowledge of any following __enter___.

The tricky thing is that you're making `PassContext::EnterWithScope` into a composite with statement--in other words, calling `EnterWithScope` calls multiple `PassInstrument::EnterPassContext`. Each `EnterPassContext` can't know whether or not another `EnterPassContext` is going to error, so doing things like opening files isn't 100% safe here--a following `EnterPassContext` which errors would leave the file open. Depending on the severity of the error, the user could retry compilation and run into a more complex, harder-to-reproduce error.

However, in every other error case (i.e. if errors occur in shouldRun, beforePass, pass, afterPass), `EnterPassContext` can count on `ExitPassContext` running. So I'd just suggest to add logic to call `ExitPassContext` while looping through and calling `EnterPassContext`. I added a comment where I think it should go.

i think if you do that, then that will button up this interface and it should handle errors in the same way as you'd expect given logic in `include/tvm/support/with.h`. You can test this by adding the `%1 exit ctx` here.

lmk if that seems unreasonable, i could be missing something. i do think that fixing the hole in the ExitPassContext is worth the potential complexity of the double-error, because at least the user is working with a simpler abstraction layer.

Reply:
Thanks for the detail explaining. I agree.